doc/ftpshut.8
.\"
.\" Copyright (c) 1999-2003 WU-FTPD Development Group. 
.\" All rights reserved.
.\" 
.\" Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994 
.\" The Regents of the University of California.  Portions Copyright (c) 
.\" 1993, 1994 Washington University in Saint Louis.  Portions Copyright 
.\" (c) 1996, 1998 Berkeley Software Design, Inc.  Portions Copyright (c) 
.\" 1998 Sendmail, Inc.  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric 
.\" P. Allman.  Portions Copyright (c) 1989 Massachusetts Institute of 
.\" Technology.  Portions Copyright (c) 1997 Stan Barber.  Portions 
.\" Copyright (C) 1991, 1992, 1993, 1994, 1995, 1996, 1997 Free Software 
.\" Foundation, Inc.  Portions Copyright (c) 1997 Kent Landfield. 
.\"
.\" Use and distribution of this software and its source code are governed 
.\" by the terms and conditions of the WU-FTPD Software License ("LICENSE"). 
.\"
.\"     $Id: ftpshut.8,v 1.9 2009/05/17 15:18:05 wmaton Exp $
.\"
.TH ftpshut 8
.SH NAME
ftpshut \- close down the ftp servers at a given time
.SH SYNTAX
.B ftpshut
[
.B \-V
] [
.B \-l
min] [
.B \-d
min]
.I time [ warning-message ... ]
.SH DESCRIPTION
The
.B ftpshut
command
provides an automated shutdown procedure that a superuser
can use to notify ftp users
when the ftp server is shutting down.
.PP
The
.B \-V
option causes the program to display copyright and version information, then
terminate.
.PP
The
.I time
is the time at which
.B ftpshut
will bring the ftp servers down. 
It may be the word
`now',
indicating an immediate shutdown,
or specify a future time in one of two formats:
.I + number
or
.I HHMM.
The first form brings the ftp servers down in
.I number
minutes.
The second brings the ftp servers down at the time of day indicated,
using a 24\-hour clock format.
.PP
Ten minutes before shutdown, or immediately if
.B ftpshut
is timed for less than ten minutes, new ftp
access will be disabled.  This time may be 
adjusted through the -l flag.
.PP
Five minutes before shutdown, or immediately if
.B ftpshut
is timed for less than five minutes, all
current ftp connections will be disconnected.
This time may be adjusted through the -d flag.
.PP
The 
.I [ warning-message ... ]
will be formatted to be 75 characters wide.
.B ftpshut
knows about the actual string length of
the magic cookies.
.PP
Running this command will create a file that the server will use to
determine when to shutdown. Once the shutdown has occured, the server
will continue to refuse connections until this file is removed. The location
of this file is specified by the 
.I shutdown
stanza in the 
.B ftpaccess
file.
.PP
The following magic cookies are available:
.nf
.ta \w'Request        'u
    \fBCookie      Description\fP
      %s      time system is going to shut down
      %r      time new connections will be denied
      %d      time current connections will be dropped
      %C      current working directory
      %E      the maintainer's email address as defined in ftpaccess
      %F      free space in partition of CWD (kbytes)
              [not currently supported on all systems]
      %L      local host name
      %M      maximum allowed number of users in this class
      %N      current number of users in this class
      %R      remote host name
      %T      local time (form Thu Nov 15 17:12:42 1990)
      %U      username given at login time
.fi
.SH RESTRICTIONS
You can kill the servers only between now and 23:59, if
you use the absolute time.
.SH SEE ALSO
.BR shutdown(1) , 
.BR ftpaccess(5) ,
.BR ftprestart(8)


CHANGES
 
  Copyright (c) 1999-2003 WU-FTPD Development Group.  
  All rights reserved.
  
  Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994
    The Regents of the University of California.
  Portions Copyright (c) 1993, 1994 Washington University in Saint Louis.
  Portions Copyright (c) 1996, 1998 Berkeley Software Design, Inc.
  Portions Copyright (c) 1989 Massachusetts Institute of Technology.
  Portions Copyright (c) 1998 Sendmail, Inc.
  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric P.  Allman.
  Portions Copyright (c) 1997 Stan Barber.
  Portions Copyright (c) 1997 Kent Landfield.
  Portions Copyright (c) 1991, 1992, 1993, 1994, 1995, 1996, 1997
    Free Software Foundation, Inc.  
 
  Use and distribution of this software and its source code are governed 
  by the terms and conditions of the WU-FTPD Software License ("LICENSE").
 
  If you did not receive a copy of the license, it may be obtained online
  at http://www.wu-ftpd.info/license.html.
 
  $Id: CHANGES,v 1.47 2016/03/14 23:57:54 wmaton Exp $

###########################################################################

Changes in 2.8.0-CC5: Released

 o  Updates to config.guess and config.sub to modernise the autoconf
    build system.

 o  configure script regenerated.

 o  Describe auto-buf tuning configure option in README.AUTOCONF.

 o  Add option (-R) to disable REST command in cases where download
    managers can abuse the server by opening multiple connections to
    retrieve chunks of the same file.  Based on a patch by Philippe Troin
    (See Debian bug ID 138497).

 o  Some minor man page cleanups.

 o  More syncs to Debian patches dating back to 2.4.2.

 o  Update INSTALL instructions.

 o  Changed the TLS default cipher from ALL:!EXP to exclude most insecure
    or unsupported ciphers within modern versions of OpenSSL.

Changes in 2.8.0-CC4: Released 7 January, 2012

 o  Thanks to a patch by Laurent CONSTANTIN, calling SSL_library_init()
    will no longer create an error within the SSL code related to a lack
    of ciphers.  The patch is applied to src/tlsutil.c.

 o  Another patch by Laurent CONSTANTIN, this time eliminating extra
    carriage returns that aren't necessary when calling syslog().

 o  Added -c option which displays the compile-time options that were used
    to build WU-FTPD. 

Changes in 2.8.0-CC3: Released 6 November, 2011

 o  Many URL updates.

 o  The default email address used by the server in absence of an email
    directive in ftpaccess is ftp-bugs, and is noted in ftpaccess now.

 o  Added code for ADLER-32 checksums for future extension to the list
    of algorithms available to do checksums.

 o  hard-coded links to cksum and md5sum within the FTP root now point to
    these utilities in /bin instead of /usr/bin.

 o  As defined in GFD.20 (GridFTP:  Protocol Extension to FTP for the
    Grid) by the Global Grid Forum, the SBUF command has been introduced.
    This calls the same code as SITE BUFSIZE would.  SITE BUFSIZE is
    still present for backwards compatibility.

 o  Thanks to a contribution by Austin Ellis, added RHEL6 packaging info
    to the codebase in packages/rpm.  This also includes a patch for
    building on that particular system.

 o  Updated ftpd.8 manpage documenting some missing SITE commands that
    had been in the server since 1998.

 o  First cut at implementing the remainder of the RFC3659 extended FTP
    commands.  FEAT, MLST, OPTS and MLSD are now supported.  Based upon
    early gridFTP releases which used WU-FTPD 2.6.0 and 2.6.1 code.

 o  According to draft-somers-ftp-mfxx-04.txt, defined the stubs for the
    MFF, MFCT and MFMT commands.  These may or may not be supported in the
    future, but at the least they are defined for now.  If so, they may
    become SITE commands instead if the spec is not approved.

 o  Updated doc/examples/ftpconversions based on Debian changes introduced
    sometime in 1995.  The server can now deal with such things as bzip2,
    xz, zip and unzip.

 o  If there's an empty /etc/fstab, a segfault can be produced.  A fix
    from Debian BugID 568235 to path_to_device in src/extensions.c addresses
    this.  Patch by Chris Butler.

 o  Added another possible path for quota.h, patch by Chris Butler of
    Debian.

 o  Added missing handler for SIGTERM, patch by Chris Butler of Debian
    related to Debian bug #536774.

 o  Changed -h invocation on the commandline to print out a very brief help
    listing all commandline options that could be supplied to the daemon.
    Also note that some switches are or have an effect on the configuration
    file.  The former function assigned to the 'h' option is now 'H'.

 o  The former function assigned to the 'h' option is now 'H'.  Said it
    twice now.  The manpage has been updated to reflect the changes.

 o  Implemented the CSID command as found in the IETF draft
    draft-peterson-streamlined-ftp-command-extensions.  This is meant
    to be a standards-based replacement for the CLNT command.  Currently
    this command is hard-wired with the server name and version, but the 
    next release will permit customisation.

###########################################################################

Changes in 2.8.0-CC2: Released 26 June, 2009

 o  Cleaned-up and updated the ftpd.8 manpage which had languished for
    some time.

 o  Added Auto tuning/auto buf, with added commands SITE BUFSIZE and
    SITE BUFSIZEMEASURE.  Use --enable-autobuf to enable this feature.
    Based on a patch on work done on WU-FTPD-2.6.1 by Gaurav Navlakha, et al
    of DAST, NLANR (now CAIDA).  This code also contains the ability to
    report back on buffer usage.  Note that an autobuf client needs to
    be used to take advantage of this feature.  For the time being, the
    code for buffer measurement is IPv4 only.

 o  Updated src/extensions.c to take into account a change in Linux quota
    structure.  Based on a patch by Chris Butler.


src/ftpshut.c
/****************************************************************************  
 
  Copyright (c) 1999-2003 WU-FTPD Development Group.  
  All rights reserved.
  
  Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994
    The Regents of the University of California.
  Portions Copyright (c) 1993, 1994 Washington University in Saint Louis.
  Portions Copyright (c) 1996, 1998 Berkeley Software Design, Inc.
  Portions Copyright (c) 1989 Massachusetts Institute of Technology.
  Portions Copyright (c) 1998 Sendmail, Inc.
  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric P.  Allman.
  Portions Copyright (c) 1997 by Stan Barber.
  Portions Copyright (c) 1997 by Kent Landfield.
  Portions Copyright (c) 1991, 1992, 1993, 1994, 1995, 1996, 1997
    Free Software Foundation, Inc.  
 
  Use and distribution of this software and its source code are governed 
  by the terms and conditions of the WU-FTPD Software License ("LICENSE").
 
  If you did not receive a copy of the license, it may be obtained online
  at http://www.wu-ftpd.info/license.html.
 
  $Id: ftpshut.c,v 1.10 2011/11/28 01:54:16 wmaton Exp $
 
****************************************************************************/
/* ftpshut 
 * ======= 
 * creates the ftpd shutdown file.
 */

#include "config.h"
#include "proto.h"

#include <errno.h>
#include <pwd.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#if defined(TIME_WITH_SYS_TIME)
#  include <time.h>
#  include <sys/time.h>
#else /* !(defined(TIME_WITH_SYS_TIME)) */ 
#  if defined(HAVE_SYS_TIME_H)
#    include <sys/time.h>
#  else /* !(defined(HAVE_SYS_TIME_H)) */ 
#    include <time.h>
#  endif /* !(defined(HAVE_SYS_TIME_H)) */ 
#endif /* !(defined(TIME_WITH_SYS_TIME)) */ 

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/file.h>
#include <sys/param.h>
#if defined(VIRTUAL) && defined(INET6)
#  include <netinet/in.h>
#endif /* defined(VIRTUAL) && defined(INET6) */ 

#include "pathnames.h"

#define WIDTH  70

int verbose = 0;
int denyoffset = 10;		/* default deny time   */
int discoffset = 5;		/* default disc time   */
char *message = "System shutdown at %s";	/* default message     */

struct tm *tp;

#define MAXVIRTUALS 512

char *progname;
char *msgfiles[MAXVIRTUALS];
int numfiles = 0;

void print_copyright(void);

static int newfile(char *fpath)
{
    int i;
    int fnd;

    /*  
       ** Check to see if the message file path has already been
       ** seen. If so then there is no need to create it again.
     */

    fnd = 0;
    for (i = 0; i < numfiles; i++) {
	if (strcmp(msgfiles[i], fpath) == 0) {
	    fnd = 1;
	    break;
	}
    }
    if (!fnd) {
	msgfiles[numfiles++] = strdup(fpath);
	return (1);
    }
    return (0);
}

static int shutdown_msgfile(char *filename, char *buffer)
{
    FILE *fp;
    mode_t oldmask;

    oldmask = umask(022);
    fp = fopen(filename, "w");
    (void) umask(oldmask);
    if (fp == NULL) {
	fprintf(stderr, "%s: could not open shutdown file %s: %s\n",
		progname, filename, strerror(errno));
	return (1);
    }

    fprintf(fp, "%.4d %.2d %.2d %.2d %.2d %.4d %.4d\n",
	    (tp->tm_year) + 1900,
	    tp->tm_mon,
	    tp->tm_mday,
	    tp->tm_hour,
	    tp->tm_min,
	    denyoffset,
	    discoffset);
    fprintf(fp, "%s\n", buffer);
    fclose(fp);
    if (verbose)
	printf("%s: %s created\n", progname, filename);
    return (0);
}

static void massage(char *buf)
{
    char *sp = NULL;
    char *ptr;
    int i = 0;
    int j = 0;

    ptr = buf;

    while (*ptr++ != '\0') {
	++i;

	/* if we have a space, keep track of where and at what "count" */

	if (*ptr == ' ') {
	    sp = ptr;
	    j = i;
	}
	/* magic cookies... */

	if (*ptr == '%') {
	    ++ptr;
	    switch (*ptr) {
	    case 'r':
	    case 's':
	    case 'd':
	    case 'T':
		i = i + 24;
		break;
	    case '\n':
		i = 0;
		break;
	    case 'C':
	    case 'R':
	    case 'L':
	    case 'U':
		i = i + 10;
		break;
	    case 'M':
	    case 'N':
		i = i + 3;
		break;
	    case '\0':
		return;
		/* break; */
	    default:
		i = i + 1;
		break;
	    }
	}
	/* break up the long lines... */

	if ((i >= WIDTH) && (sp != NULL)) {
	    *sp = '\n';
	    sp = NULL;
	    i = i - j;
	}
    }
}

static void usage(int exitval)
{
    fprintf(stderr,
	    "Usage: %s [-d min] [-l min] now [\"message\"]\n", progname);
    fprintf(stderr,
	    "       %s [-d min] [-l min] +dd [\"message\"]\n", progname);
    fprintf(stderr,
	    "       %s [-d min] [-l min] HHMM [\"message\"]\n", progname);
    exit(exitval);
}

int main(int argc, char **argv)
{
    time_t c_time = 0;

    char buf[BUFSIZ];

    int c;
    extern int optind;
    extern char *optarg;

    FILE *accessfile;
    char *aclbuf, *myaclbuf, *crptr;
    char *sp = NULL;
    char linebuf[1024];
    char shutmsg[BUFSIZ];
    char anonpath[MAXPATHLEN];
    struct stat finfo;
    struct passwd *pwent;

#if defined(VIRTUAL)
    char *cp = NULL;
    FILE *svrfp;
#  if defined(INET6)
    char hostaddress[INET6_ADDRSTRLEN];
#  else /* !(defined(INET6)) */ 
    char hostaddress[32];
#  endif /* !(defined(INET6)) */ 
    char root[MAXPATHLEN];
    char accesspath[MAXPATHLEN];
    char configdir[MAXPATHLEN];
    char altmsgpath[MAXPATHLEN];
#endif /* defined(VIRTUAL) */ 

    if ((progname = strrchr(argv[0], '/')))
	++progname;
    else
	progname = argv[0];

    while ((c = getopt(argc, argv, "vVl:d:")) != EOF) {
	switch (c) {
	case 'v':
	    verbose++;
	    break;
	case 'l':
	    denyoffset = atoi(optarg);
	    break;
	case 'd':
	    discoffset = atoi(optarg);
	    break;
	case 'V':
	    print_copyright();
	    exit(0);
	default:
	    usage(-1);
	}
    }

    if ((accessfile = fopen(_PATH_FTPACCESS, "r")) == NULL) {
	if (errno != ENOENT)
	    fprintf(stderr, "%s: could not open access file %s: %s\n",
		    progname, _PATH_FTPACCESS, strerror(errno));
	exit(1);
    }
    if (fstat(fileno(accessfile), &finfo) != 0) {
	fprintf(stderr, "%s: could not fstat() access file %s: %s\n",
		progname, _PATH_FTPACCESS, strerror(errno));
	exit(1);
    }
    if (finfo.st_size == 0) {
	fprintf(stderr, "%s: no shutdown path defined in ftpaccess file %s.\n",
		progname, _PATH_FTPACCESS);
	exit(1);
    }
    else {
	if (!(aclbuf = (char *) malloc(finfo.st_size + 1))) {
	    fprintf(stderr, "%s: could not malloc aclbuf: %s\n",
		    progname, strerror(errno));
	    exit(1);
	}
	(void) fread(aclbuf, finfo.st_size, 1, accessfile);
	*(aclbuf + finfo.st_size) = '\0';
    }

    myaclbuf = aclbuf;
    while (*myaclbuf != '\0') {
	if (strncasecmp(myaclbuf, "shutdown", 8) == 0) {
	    for (crptr = myaclbuf; *crptr++ != '\n';);
	    *--crptr = '\0';
	    strlcpy(linebuf, myaclbuf, sizeof(linebuf));
	    *crptr = '\n';
	    (void) strtok(linebuf, " \t");	/* returns "shutdown" */
	    sp = strtok(NULL, " \t");	/* returns shutdown path */
	    strlcpy(shutmsg, sp, sizeof(shutmsg)); /* save for future use */
	}
	while (*myaclbuf && *myaclbuf++ != '\n');
    }

    /* three cases 
     * -- now 
     * -- +ddd 
     * -- HHMM 
     */

    c = -1;

    if (optind < argc) {
	if (!strcasecmp(argv[optind], "now")) {
	    c_time = time(0);
	    tp = localtime(&c_time);
	}
	else if ((*(argv[optind])) == '+') {
	    c_time = time(0);
	    c_time += 60 * atoi(++(argv[optind]));
	    tp = localtime(&c_time);
	}
	else if ((c = atoi(argv[optind])) >= 0) {
	    c_time = time(0);
	    tp = localtime(&c_time);
	    tp->tm_hour = c / 100;
	    tp->tm_min = c % 100;

	    if ((tp->tm_hour > 23) || (tp->tm_min > 59)) {
		fprintf(stderr, "%s: illegal time format.\n", progname);
		exit(1);
	    }
	}
    }
    if (c_time <= 0) {
	usage(1);
    }

    if (sp == NULL) {
	fprintf(stderr, "%s: no shutdown path defined in ftpaccess file %s.\n",
		progname, _PATH_FTPACCESS);
	exit(1);
    }

    /* do we have a shutdown message? */
    if (++optind < argc) 
	strlcpy(buf, argv[optind++], sizeof(buf));
    else
	strlcpy(buf, message, sizeof(buf));

    massage(buf);

    /* 
       ** Create the system shutdown message file at the location
       ** specified in the ftpaccess 'shutdown' directive.  This
       ** is for support of real system users.
     */
    c = shutdown_msgfile(shutmsg, buf);
    msgfiles[numfiles++] = shutmsg;

    /* 
       ** Determine if the site supports anonymous ftp and if so, create
       ** the shutdown message file in the anonymous ftp area as well
       ** so that shutdown works appropriately for both real and guest 
       ** accounts. Save in msgfiles array for later comparison.
     */

    if ((pwent = getpwnam("ftp")) != NULL) {
	snprintf(anonpath, sizeof(anonpath), "%s%s", pwent->pw_dir, shutmsg);
	if (newfile(anonpath))
	    c += shutdown_msgfile(anonpath, buf);
    }

#if defined(VIRTUAL)
    /*
       ** Search the Master access file for virtual ftp servers.
       ** If found, construct a path to the shutdown message file
       ** under the virtual server's root.  Don't duplicate what
       ** is specified in the "ftp" account directory information.
     */

    rewind(accessfile);

    while (fgets(linebuf, sizeof(linebuf) - 1, accessfile) != NULL) {
	if (strncasecmp(linebuf, "virtual", 7) == 0) {

	    if ((sp = strstr(linebuf, "root")) != NULL) {
		if ((cp = strchr(sp, '\n')) != NULL)
		    *cp = '\0';	/* strip newline */

		sp += 4;	/* skip past "root" keyword */

		while (*sp && isspace(*sp))	/* skip whitespace to root path */
		    sp++;
		cp = sp;
		while (*sp && !isspace(*sp))
		    sp++;
		*sp = '\0';	/* truncate blanks, comments etc. */

		snprintf(altmsgpath, sizeof(altmsgpath), "%s%s", cp, shutmsg);

		if (newfile(altmsgpath))
		    c += shutdown_msgfile(altmsgpath, buf);
	    }
	}
    }

    /*
       ** Need to deal with the access files at the virtual domain directory
       ** locations specified in the ftpservers file. 
     */

    if ((svrfp = fopen(_PATH_FTPSERVERS, "r")) != NULL) {
	while (read_servers_line(svrfp, hostaddress, sizeof(hostaddress), configdir, sizeof(configdir)) == 1) {
	    /* get rid of any trailing slash */
	    sp = configdir + (strlen(configdir) - 1);
	    if (*sp == '/')
		*sp = '\0';

	    /* 
	       ** check to see that a valid directory value was
	       ** supplied and not something such as "INTERNAL"
	       **
	       ** It is valid to have a string such as "INTERNAL" in the
	       ** ftpservers entry. This is not an error. Silently ignore it.
	     */

	    if ((stat(configdir, &finfo) < 0) ||
		((finfo.st_mode & S_IFMT) != S_IFDIR))
		continue;

	    snprintf(accesspath, sizeof(accesspath), "%s/ftpaccess", configdir);

	    (void) fclose(accessfile);

	    if ((accessfile = fopen(accesspath, "r")) == NULL) {
		if (errno != ENOENT) {
		    fprintf(stderr, "%s: could not open access file %s: %s\n",
			    progname, accesspath, strerror(errno));
		    continue;
		}
	    }

	    /* need to find the root path */

	    while (fgets(linebuf, sizeof(linebuf) - 1, accessfile) != NULL) {
		if ((sp = strstr(linebuf, "root")) != NULL) {
		    if ((cp = strchr(sp, '\n')) != NULL)
			*cp = '\0';	/* strip newline */
		    sp += 4;	/* skip past "root" keyword */

		    while (*sp && isspace(*sp))		/* skip whitespace to path */
			sp++;
		    cp = sp;
		    while (*sp && !isspace(*sp))
			sp++;
		    *sp = '\0';	/* truncate blanks, comments etc. */
		    strlcpy(root, cp, sizeof(root));
		    break;
		}
	    }
	    /* need to find the shutdown message file path */

	    rewind(accessfile);

	    while (fgets(linebuf, sizeof(linebuf) - 1, accessfile) != NULL) {
		if ((sp = strstr(linebuf, "shutdown")) != NULL) {
		    if ((cp = strchr(sp, '\n')) != NULL)
			*cp = '\0';	/* strip newline */
		    sp += 8;	/* skip past "root" keyword */

		    while (*sp && isspace(*sp))		/* skip whitespace to path */
			sp++;
		    cp = sp;
		    while (*sp && !isspace(*sp))
			sp++;
		    *sp = '\0';	/* truncate blanks, comments etc. */
		    break;
		}
	    }

	    /*
	       ** check to make sure the admin hasn't specified 
	       ** a complete path in the 'shutdown' directive.
	     */
	    if ((sp = strstr(cp, root)) == NULL)
		snprintf(altmsgpath, sizeof(altmsgpath), "%s%s", root, cp);

	    /*
	       ** Check to see if the message file has been created elsewhere.
	     */
	    if (newfile(altmsgpath))
		c += shutdown_msgfile(altmsgpath, buf);
	}
	fclose(svrfp);
    }
#endif /* defined(VIRTUAL) */ 

    fclose(accessfile);
    free(aclbuf);
    exit(c > 0 ? 1 : 0);
}
